src/vs/workbench/api/common/extHostTelemetry.ts

```diff
@@ -3,49 +3,204 @@
 		this._oldTelemetryEnablement = this.getTelemetryConfiguration();
 		this._level = level;
+		const telemetryDetails = this.getTelemetryDetails();
+
+		// Loop through all loggers and update their level
+		this._telemetryLoggers.forEach(logger => {
+			logger.updateTelemetryEnablements(telemetryDetails.isUsageEnabled, telemetryDetails.isErrorsEnabled);
+		});
+
 		if (this._oldTelemetryEnablement !== this.getTelemetryConfiguration()) {
 			this._onDidChangeTelemetryEnabled.fire(this.getTelemetryConfiguration());
 		}
 		this._onDidChangeTelemetryConfiguration.fire(this.getTelemetryDetails());
 	}
 }
 
+export class ExtHostTelemetryLogger {
+	private _appender: vscode.TelemetryAppender;
+	private readonly _onDidChangeEnableStates = new Emitter<vscode.TelemetryLogger>();
+	private _telemetryEnablements: { isUsageEnabled: boolean; isErrorsEnabled: boolean };
+	private _apiObject: vscode.TelemetryLogger | undefined;
+	constructor(
+		appender: vscode.TelemetryAppender,
+		private readonly _extension: IExtensionDescription,
+		private readonly _logger: ILogger,
+		private readonly _commonProperties: Record<string, any>,
+		telemetryEnablements: { isUsageEnabled: boolean; isErrorsEnabled: boolean }) {
+		this._appender = appender;
+		this._telemetryEnablements = { isUsageEnabled: telemetryEnablements.isUsageEnabled, isErrorsEnabled: telemetryEnablements.isErrorsEnabled };
+	}
+
+	updateTelemetryEnablements(isUsageEnabled: boolean, isErrorsEnabled: boolean): void {
+		if (this._apiObject) {
+			this._telemetryEnablements = { isUsageEnabled, isErrorsEnabled };
+			this._onDidChangeEnableStates.fire(this._apiObject);
+		}
+	}
+
+	mixInCommonPropsAndCleanData(data: Record<string, any>): Record<string, any> {
+		if (!this._appender.ignoreBuiltInCommonProperties) {
+			data = mixin(data, this._commonProperties);
+		}
+		if (this._appender.additionalCommonProperties) {
+			data = mixin(data, this._appender.additionalCommonProperties);
+		}
+
+		data = cleanData(data, []);
+
+		return data;
+	}
+
+	private logEvent(eventName: string, data?: Record<string, any>): void {
+		eventName = this._extension.identifier.value + '/' + eventName;
+		data = this.mixInCommonPropsAndCleanData(data || {});
+		this._appender.logEvent(eventName, data);
+		this._logger.trace(eventName, data);
+	}
+
+	logUsage(eventName: string, data?: Record<string, any>): void {
+		if (!this._telemetryEnablements.isUsageEnabled) {
+			return;
+		}
+		this.logEvent(eventName, data);
+	}
+
+	logError(eventNameOrException: Error | string, data?: Record<string, any>): void {
+		if (!this._telemetryEnablements.isErrorsEnabled) {
+			return;
+		}
+		if (typeof eventNameOrException === 'string') {
+			this.logEvent(eventNameOrException, data);
+		} else {
+			// TODO @lramos15, implement cleaning for and logging for this case
+			this._appender.logException(eventNameOrException, data);
+		}
+	}
+
+	get apiTelemetryLogger(): vscode.TelemetryLogger {
```

Comment:
You could use the `Lazy` helper here

Reply:
Not too familiar with the lazy helper, what exactly does that do here?